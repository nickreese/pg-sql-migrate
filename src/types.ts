import { ClientConfig, ClientBase } from 'pg';

export interface Config {
  client: ClientConfig | string;
  dir: string;
  table: string;
}

export interface Migration {
  id: string;
  name: string;
  content: string;
}

export interface PGClient {
  query: ClientBase['query'];
}

export const DEFAULT_CONFIG_FILE = 'pg-sql-migrate.config.json';

export const CONFIG_DEFAULTS = {
  dir: 'migrations',
  table: 'migrations',
};
