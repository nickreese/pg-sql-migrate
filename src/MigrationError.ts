import { Migration } from './types';

export class MigrationError extends Error {
  public migration: Migration;
  public constructor(message: string | undefined, migration: Migration) {
    super(message);
    this.migration = migration;
  }
}
