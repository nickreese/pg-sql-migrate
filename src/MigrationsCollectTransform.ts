import { Transform } from 'stream';
import { Migration } from './types';

export class MigrationsCollectTransform extends Transform {
  public migrations: Migration[] = [];
  public constructor() {
    super({ objectMode: true });
  }

  public async _transform(
    data: Migration,
    encoding: string,
    callback: (error: Error | null | undefined, data: Migration) => void,
  ): Promise<void> {
    this.migrations.push(data);
    callback(null, data);
  }
}
