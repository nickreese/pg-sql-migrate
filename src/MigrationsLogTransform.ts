import { Transform, TransformCallback } from 'stream';
import { Migration } from './types';

export class MigrationsLogTransform extends Transform {
  public constructor() {
    super({ objectMode: true });
  }

  public _transform(data: Migration, encoding: string, callback: TransformCallback): void {
    console.log(`[${data.id}] ${data.name}`);
    callback(null, data);
  }
}
