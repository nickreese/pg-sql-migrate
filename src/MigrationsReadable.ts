import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { QueryResult } from 'pg';
import { Readable } from 'stream';
import { Migration, PGClient } from './types';

export const nameParts = (name: string): string[] => {
  const [first, ...rest] = name.split('_');
  return [first, rest.join('_')];
};

export class MigrationsReadable extends Readable {
  private current: number = 0;
  private migrationFiles?: string[];
  private pg: PGClient;
  private table: string;
  private directory: string;

  public constructor(pg: PGClient, table: string, directory: string) {
    super({ objectMode: true });
    this.pg = pg;
    this.table = table;
    this.directory = directory;
  }

  private async initialize(): Promise<void> {
    const migrationFiles = readdirSync(this.directory).filter(file => file.endsWith('.pgsql'));
    await this.initState();
    const completed = await this.loadState();

    this.migrationFiles = migrationFiles.filter(file => !completed.includes(nameParts(file)[0]));
  }

  private async next(): Promise<Migration | null> {
    if (!this.migrationFiles) {
      await this.initialize();
    }

    if (this.migrationFiles && this.migrationFiles[this.current]) {
      const file = this.migrationFiles[this.current++];
      const [id, name] = nameParts(file);
      const content = readFileSync(join(this.directory, file)).toString();
      const migration: Migration = { id, name, content };
      return migration;
    } else {
      return null;
    }
  }

  public async _read(): Promise<void> {
    this.push(await this.next());
  }

  private async initState(): Promise<QueryResult> {
    return await this.pg.query(`CREATE TABLE IF NOT EXISTS ${this.table} (id VARCHAR PRIMARY KEY)`);
  }

  private async loadState(): Promise<string[]> {
    return (await this.pg.query(`SELECT id FROM ${this.table}`)).rows.map(row => row.id);
  }
}
