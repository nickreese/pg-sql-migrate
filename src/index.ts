export { MigrationError } from './MigrationError';
export { MigrationsReadable } from './MigrationsReadable';
export { MigrationsWritable } from './MigrationsWritable';
export { MigrationsCollectTransform } from './MigrationsCollectTransform';
export { migrate, executeMigrations, loadConfig } from './migrate';
export { Migration, Config, DEFAULT_CONFIG_FILE, CONFIG_DEFAULTS } from './types';
