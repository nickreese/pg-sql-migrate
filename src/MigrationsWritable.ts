import { Writable } from 'stream';
import { MigrationError } from './MigrationError';
import { Migration, PGClient } from './types';

export class MigrationsWritable extends Writable {
  private pg: PGClient;
  private table: string;
  public constructor(pg: PGClient, table: string) {
    super({ objectMode: true });
    this.pg = pg;
    this.table = table;
  }

  public async _write(
    migration: Migration,
    encoding: string,
    callback: (error?: Error | null) => void,
  ): Promise<void> {
    try {
      await this.pg.query(migration.content);
      await this.pg.query(`INSERT INTO ${this.table} VALUES ($1)`, [migration.id]);
      callback(null);
    } catch (error) {
      callback(new MigrationError(error.message, migration));
    }
  }
}
