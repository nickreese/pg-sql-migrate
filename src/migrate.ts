import { loadConfigFile } from '@ovotech/config-file';
import { Client } from 'pg';
import { MigrationsReadable, MigrationsWritable } from './';
import { Config, CONFIG_DEFAULTS, DEFAULT_CONFIG_FILE, Migration, PGClient } from './types';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { MigrationsCollectTransform } from './MigrationsCollectTransform';

export const executeMigrations = async (
  pg: PGClient,
  table: string,
  directory: string,
): Promise<Migration[]> => {
  const read = new MigrationsReadable(pg, table, directory);
  const sink = new MigrationsWritable(pg, table);
  const collect = new MigrationsCollectTransform();

  await promisify(pipeline)(read, collect, sink);
  return collect.migrations;
};

export const loadConfig = (file = DEFAULT_CONFIG_FILE, env = process.env): Config =>
  loadConfigFile<Config>({ file, env, defaults: CONFIG_DEFAULTS, required: ['client'] });

export const migrate = async (
  config?: Partial<Config> | string,
  env = process.env,
): Promise<Migration[]> => {
  const { client, table, directory } =
    typeof config === 'object' ? { ...CONFIG_DEFAULTS, ...config } : loadConfig(config, env);

  const pg = new Client(client);
  await pg.connect();
  const results = await executeMigrations(pg, table, directory);
  await pg.end();
  return results;
};
