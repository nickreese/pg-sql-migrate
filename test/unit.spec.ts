import { join } from 'path';
import { executeMigrations, MigrationsReadable, MigrationsWritable } from '../src';

describe('Unit test', () => {
  it('Should use streams to execute migrations', async () => {
    const pg = { query: jest.fn() };

    pg.query
      .mockResolvedValueOnce({})
      .mockResolvedValueOnce({
        rows: [{ id: '2018-12-31T11:57:10.022Z' }, { id: '2018-12-31T12:10:49.562Z' }],
      })
      .mockResolvedValue({});

    const readable = new MigrationsReadable(pg, 'test1', join(__dirname, 'migrations'));
    const writable = new MigrationsWritable(pg, 'test1');

    readable.pipe(writable);

    await new Promise(resolve => writable.on('finish', resolve));

    expect(pg.query).toBeCalledTimes(6);

    expect(pg.query).toHaveBeenNthCalledWith(
      1,
      'CREATE TABLE IF NOT EXISTS test1 (id VARCHAR PRIMARY KEY)',
    );
    expect(pg.query).toHaveBeenNthCalledWith(2, 'SELECT id FROM test1');
    expect(pg.query).toHaveBeenNthCalledWith(
      3,
      'CREATE TABLE my_test (id INTEGER PRIMARY KEY, name VARCHAR);\n',
    );
    expect(pg.query).toHaveBeenNthCalledWith(4, 'INSERT INTO test1 VALUES ($1)', [
      '2018-12-31T11:12:39.672Z',
    ]);
    expect(pg.query).toHaveBeenNthCalledWith(
      5,
      'ALTER TABLE my_test ADD COLUMN additional_3 VARCHAR;\n',
    );
    expect(pg.query).toHaveBeenNthCalledWith(6, 'INSERT INTO test1 VALUES ($1)', [
      '2019-01-02T08:36:08.858Z',
    ]);
  });

  it('Should await migrate function to run streams', async () => {
    const pg = { query: jest.fn() };

    pg.query
      .mockResolvedValueOnce({})
      .mockResolvedValueOnce({
        rows: [{ id: '2018-12-31T12:10:49.562Z' }, { id: '2019-01-02T08:36:08.858Z' }],
      })
      .mockResolvedValue({});

    const result = await executeMigrations(pg, 'testing', join(__dirname, 'migrations'));
    expect(pg.query).toBeCalledTimes(6);

    expect(pg.query).toHaveBeenNthCalledWith(
      1,
      'CREATE TABLE IF NOT EXISTS testing (id VARCHAR PRIMARY KEY)',
    );
    expect(pg.query).toHaveBeenNthCalledWith(2, 'SELECT id FROM testing');
    expect(pg.query).toHaveBeenNthCalledWith(
      3,
      'CREATE TABLE my_test (id INTEGER PRIMARY KEY, name VARCHAR);\n',
    );
    expect(pg.query).toHaveBeenNthCalledWith(4, 'INSERT INTO testing VALUES ($1)', [
      '2018-12-31T11:12:39.672Z',
    ]);
    expect(pg.query).toHaveBeenNthCalledWith(
      5,
      'ALTER TABLE my_test ADD COLUMN additional VARCHAR;\n',
    );
    expect(pg.query).toHaveBeenNthCalledWith(6, 'INSERT INTO testing VALUES ($1)', [
      '2018-12-31T11:57:10.022Z',
    ]);

    expect(result).toMatchSnapshot();
  });
});
