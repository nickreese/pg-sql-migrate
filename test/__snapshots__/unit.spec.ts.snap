// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test Should await migrate function to run streams 1`] = `
Array [
  Object {
    "content": "CREATE TABLE my_test (id INTEGER PRIMARY KEY, name VARCHAR);
",
    "id": "2018-12-31T11:12:39.672Z",
    "name": "test-things.pgsql",
  },
  Object {
    "content": "ALTER TABLE my_test ADD COLUMN additional VARCHAR;
",
    "id": "2018-12-31T11:57:10.022Z",
    "name": "test-things2.pgsql",
  },
]
`;
